File: app/src/main/java/me/devsaki/hentoid/core/EmergencyRestartHandler.java
package me.devsaki.hentoid.core;

import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;

import android.content.Context;
import android.content.Intent;
import android.os.Process;

import androidx.annotation.NonNull;

import me.devsaki.hentoid.util.Helper;
import timber.log.Timber;

public class EmergencyRestartHandler implements
        java.lang.Thread.UncaughtExceptionHandler {
    private final Context myContext;
    private final Class<?> myActivityClass;

    public EmergencyRestartHandler(Context context, Class<?> c) {
        myContext = context;
        myActivityClass = c;
    }

    public void uncaughtException(@NonNull Thread thread, @NonNull Throwable exception) {
        Timber.e(exception);

        // Log the exception
        Timber.i("Logging crash exception");
        try {
            Helper.logException(exception);
        } finally {
            // Restart the Activity
            Timber.i("Restart %s", myActivityClass.getSimpleName());
            Intent intent = new Intent(myContext, myActivityClass);
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | FLAG_ACTIVITY_NEW_TASK);
            myContext.startActivity(intent);
        }

        Timber.i("Kill current process");
        Process.killProcess(Process.myPid());
        System.exit(0);
    }
}

File: app/src/main/java/me/devsaki/hentoid/core/EmergencyRestartHandler.kt
package me.devsaki.hentoid.core

import android.content.Context
import android.content.Intent
import android.os.Process
import me.devsaki.hentoid.util.logException
import timber.log.Timber
import kotlin.system.exitProcess

class EmergencyRestartHandler(val context: Context, private val myActivityClass: Class<*>) :
    Thread.UncaughtExceptionHandler {

    override fun uncaughtException(t: Thread, e: Throwable) {
        Timber.e(e)

        // Log the exception
        Timber.i("Logging crash exception")
        try {
            logException(e)
        } finally {
            // Restart the Activity
            Timber.i("Restart %s", myActivityClass.simpleName)
            val intent = Intent(context, myActivityClass)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK
            context.startActivity(intent)
        }
        Timber.i("Kill current process")
        Process.killProcess(Process.myPid())
        exitProcess(0)
    }
}